numba/cuda/cudadecl.py

```diff
@@ -276,71 +276,36 @@ def generic(self, args, kws):
         return signature(a, test, a, a)
 
 
-@register
-class Cuda_atomic_add(AbstractTemplate):
-    key = cuda.atomic.add
-
-    def generic(self, args, kws):
-        assert not kws
-        ary, idx, val = args
-
-        if ary.ndim == 1:
-            return signature(ary.dtype, ary, types.intp, ary.dtype)
-        elif ary.ndim > 1:
-            return signature(ary.dtype, ary, idx, ary.dtype)
-
-
-@register
-class Cuda_atomic_sub(AbstractTemplate):
-    key = cuda.atomic.sub
-
-    def generic(self, args, kws):
-        assert not kws
-        ary, idx, val = args
-
-        if ary.ndim == 1:
-            return signature(ary.dtype, ary, types.intp, ary.dtype)
-        elif ary.ndim > 1:
-            return signature(ary.dtype, ary, idx, ary.dtype)
-
-
-class Cuda_atomic_maxmin(AbstractTemplate):
-    def generic(self, args, kws):
-        assert not kws
-        ary, idx, val = args
-        # Implementation presently supports:
-        # float64, float32, int32, int64, uint32, uint64 only,
-        # so fail typing otherwise
-        supported_types = (types.float64, types.float32,
-                           types.int32, types.uint32,
-                           types.int64, types.uint64)
-        if ary.dtype not in supported_types:
-            return
-
-        if ary.ndim == 1:
-            return signature(ary.dtype, ary, types.intp, ary.dtype)
-        elif ary.ndim > 1:
-            return signature(ary.dtype, ary, idx, ary.dtype)
-
-
-@register
-class Cuda_atomic_max(Cuda_atomic_maxmin):
-    key = cuda.atomic.max
-
-
-@register
-class Cuda_atomic_min(Cuda_atomic_maxmin):
-    key = cuda.atomic.min
-
-
-@register
-class Cuda_atomic_nanmax(Cuda_atomic_maxmin):
-    key = cuda.atomic.nanmax
-
-
-@register
-class Cuda_atomic_nanmin(Cuda_atomic_maxmin):
-    key = cuda.atomic.nanmin
+# generate atomic operations
+def _gen(l_key, supported_types):
+    @register
+    class Cuda_atomic(AbstractTemplate):
+        key = l_key
+
+        def generic(self, args, kws):
+            assert not kws
+            ary, idx, val = args
+
+            if ary.dtype not in supported_types:
+                return
+
+            if ary.ndim == 1:
+                return signature(ary.dtype, ary, types.intp, ary.dtype)
+            elif ary.ndim > 1:
+                return signature(ary.dtype, ary, idx, ary.dtype)
+    return Cuda_atomic
+
+
+all_numba_types = (types.float64, types.float32,
+                   types.int32, types.uint32,
+                   types.int64, types.uint64)
```

Comment:
This is applying additional type restriction to the types accepted by `atomic.add` and `atomic.sub`, I think this is probably a good thing as e.g.:
```python
from numba import cuda
import numpy as np

@cuda.jit
def foo(result, values):
    tid = cuda.threadIdx.x
    bid = cuda.blockIdx.x
    bdim = cuda.blockDim.x
    i = (bid * bdim) + tid
    cuda.atomic.add(result, 0, values[i])


arr = np.ones(1, dtype=np.int8)
result = np.ones(1, dtype=np.int8)

foo[1, 1](result, arr)
print(result[0])
```
on mainline results in:
```
numba.cuda.cudadrv.error.NvvmError: Failed to compile

(0) Error: unsupported operation

NVVM_ERROR_COMPILATION
```
this also matches up with the lowering available.

CC @gmarkall

Reply:
I agree, this is a good thing.